Initial support create relationship in cost planner

pontusmelke opened the pull request:
Note that this does a little more than what is proposed to be the 3.0 functionality of `CREATE`, however removing any unwanted features once the feature set has been decided on is more or less trivial. Thus, imho, we do not need to wait for the final decision before merging this.

pontusmelke committed:
Support for create ()-[:R]->() in cost planner

pontusmelke committed:
Added support for CREATE ()-[:R1]->()-[:R2]-...->()

Commit:
Changed to stack-safe implementation of PipeExecutionPlanBuilder

pontusmelke committed:
Added scalacheck test for semantic verification of CREATE

pontusmelke committed:
Fixed bug with loops, CREATE (a:L)-[:R]->(a)

systay commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/execution/PipeExecutionPlanBuilder.scala:
```diff
@@ -23,271 +23,381 @@ import org.neo4j.cypher.internal.compiler.v3_0.ast.convert.commands.ExpressionCo
-          val source = new SingleRowPipe()
-          val ep = entityProducerFactory.nodeStartItems((planContext, StatementConverters.StartItemConverter(hint).asCommandStartItem))
-          NodeStartPipe(source, id.name, ep, Effects(ReadsAllNodes))()
+    // TODO: Maybe we shouldn't encode distinct as an empty aggregation.
```
There are a couple of TODOs left in this file that probably should not be there when we merge this in

systay replied:
I'm thinking of the TODO on line 280 and 331, not the one on 279

pontusmelke committed:
Clean up, removed inline pipe rewriter.

Instead of rewriting the pipe after it has been added to the stack
we create the correct pipe before adding it to the stack.
